spec/models/api_wrappers/kraken_api_wrapper_spec.rb

```diff
@@ -16,63 +14,148 @@
     BitexBot::Robot.setup
   end
 
-  def stub_public_client
-    api_client.stub(public: double)
+  def stub_request_helper(method:, path: '', status: 200, result: {}, error: [], header_params: {})
+    stub_request(method, "https://api.kraken.com/0#{path}")
+      .with(headers: { 'User-Agent': BitexBot.user_agent }.merge(header_params))
+      .to_return(
+        status: status,
+        body: { error: error, result: result }.to_json,
+        headers: { 'Content-Type': 'application/json' }
+      )
   end
 
-  def stub_private_client
-    api_client.stub(private: double)
+  def stub_assets
+    stub_request_helper(
+      method: :get,
+      path: '/public/AssetPairs',
+      result: {
+        XXBTZUSD:  {
+          altname: 'XBTUSD',
+          aclass_base: 'currency',
+          base: 'XXBT',
+          aclass_quote: 'currency',
+          quote: 'ZUSD',
+          lot: 'unit',
+          pair_decimals: 1,
+          lot_decimals: 8,
+          lot_multiplier: 1,
+          leverage_buy: [2, 3, 4, 5],
+          leverage_sell: [2, 3, 4, 5],
+          fees: [
+            [0, 0.26],
+            [50_000, 0.24],
+            [100_000, 0.22],
+            [250_000, 0.2],
+            [500_000, 0.18],
+            [1_000_000, 0.16],
+            [2_500_000, 0.14],
+            [5_000_000, 0.12],
+            [10_000_000, 0.1]
+          ],
+          fees_maker: [
+            [0, 0.16],
+            [50_000, 0.14],
+            [100_000, 0.12],
+            [250_000, 0.1],
+            [500_000, 0.08],
+            [1_000_000, 0.06],
+            [2_500_000, 0.04],
+            [5_000_000, 0.02],
+            [10_000_000, 0]
+          ],
+          fee_volume_currency: 'ZUSD',
+          margin_call: 80,
+          margin_stop: 40
+        }
+      }
+    )
   end
 
   it 'Sends User-Agent header' do
-    url = 'https://api.kraken.com/0/public/Depth?pair=XBTUSD'
-    stub_stuff = stub_request(:get, url).with(headers: { 'User-Agent': BitexBot.user_agent })
+    stub_assets
+    stub_stuff = stub_order_book
 
     # We don't care about the response
-    KrakenApiWrapper.order_book rescue nil
+    api_wrapper.order_book
 
-    expect(stub_stuff).to have_been_requested
+    stub_stuff.should have_been_requested
   end
 
   def stub_balance
-    api_client.private.stub(account_info: [{ taker_fees: '89.2' }])
-    api_client.private.stub(:balance) do
-      { 'XXBT': '1433.0939', 'ZUSD': '1230.0233', 'XETH': '99.7497224800' }.with_indifferent_access
-    end
+    stub_request_helper(
+      method: :post,
+      path: '/private/Balance',
+      header_params: { 'Api-Key': api_wrapper.settings.api_key },
+      result: { XXBT: '1433.0939', ZUSD: '1230.0233', ETH: '99.7497224800' }
+    )
+  end
+
+  def stub_open_orders
```

Comment:
What's the difference between this and `stub_orders`?

Reply:
none. was removed. 👍